R/tokenization.R

```diff
@@ -13,4 +13,478 @@
+
+
+#' Split text on punctuation.
+#'
+#' (R implementation of BasicTokenizer._run_split_on_punc from
+#' BERT: tokenization.py.)
+#'
+#' @param text A character scalar, encoded as utf-8.
+#'
+#' @return The input text as a character vector, split on punctuation
+#' characters.
+#'
+#' @keywords internal
+.split_on_punc <- function(text) {
+    # this feels icky, but try to break it :-P
+    # Put a unique marker around each punctuation char, then split on the
+    # marker (since we want the punctuation to be included in split).
+    sep_marker <- "a!b"
+    output <- .apply_to_chars(text,
+                              function(char) {
+                                  if(.is_punctuation(char)) {
+                                      return(paste0(sep_marker,
+                                                    char,
+                                                    sep_marker))
+                                  }
+                                  return(char)
+                              })
+    return(
+        unlist(
+            stringi::stri_split_fixed(output, sep_marker, omit_empty = TRUE)
+        )
+    )
+}
+
+
+# .tokenize_chinese_chars --------------------------------------------------
+
+
+#' Add whitespace around any CJK character.
+#'
+#' R implementation of BasicTokenizer._tokenize_chinese_chars from
+#' BERT: tokenization.py. This may result in doubled-up spaces,
+#' but that's the behavior of the Python code.
+#'
+#' @param text A character scalar.
+#'
+#' @return Text with spaces around CJK characters.
+#'
+#' @keywords internal
+.tokenize_chinese_chars <- function(text) {
+    return(
+        .apply_to_chars(text,
+                        function(char) {
+                            cp <- utf8ToInt(char)
+                            if(.is_chinese_char(cp)) {
+                                return(paste0(" ", char, " ")) #nocov
+                            }
+                            return(char)
+                        })
+    )
+}
+
+
+# .is_chinese_char ---------------------------------------------------------
+
+
+#' Check whether cp is the codepoint of a CJK character.
+#'
+#' R implementation of BasicTokenizer._is_chinese_char from
+#' BERT: tokenization.py. From that file:
+#'  This defines a "chinese character" as anything in the CJK Unicode block:
+#'   https://en.wikipedia.org/wiki/CJK_Unified_Ideographs_(Unicode_block)
+#'
+#' Note that the CJK Unicode block is NOT all Japanese and Korean characters,
+#' despite its name. The modern Korean Hangul alphabet is a different block,
+#' as is Japanese Hiragana and Katakana. Those alphabets are used to write
+#' space-separated words, so they are not treated specially and are handled
+#' like the alphabets of the other languages.
+#'
+#' @param cp A unicode codepoint, as an integer.
+#'
+#' @return Logical TRUE if cp is codepoint of a CJK character.
+#'
+#' @keywords internal
+.is_chinese_char <- function(cp) {
```

Comment:
For clarity, can you rename this to `.is_chinese_char_cp` (or pass a character in instead)? It was slightly confusing reading the tests where every other `.is_*_char` function takes a character as the input.

Reply:
good idea